File: sqlite-android-studio-plugin/src/main/java/com.alecstrong.sqlite.android/util/SqlitePsiUtils.java
package com.alecstrong.sqlite.android.util;

import com.alecstrong.sqlite.android.lang.SqliteLanguage;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFileFactory;
import com.intellij.psi.impl.PsiFileFactoryImpl;
import com.intellij.psi.tree.IElementType;
import com.intellij.psi.util.PsiTreeUtil;

@SuppressWarnings("SimplifiableIfStatement")
public class SqlitePsiUtils {
  public static PsiElement createLeafFromText(Project project, PsiElement context,
      String text, IElementType type) {
    PsiFileFactoryImpl factory = (PsiFileFactoryImpl) PsiFileFactory.getInstance(project);
    PsiElement el = factory.createElementFromText(text,
        SqliteLanguage.INSTANCE,
        type,
        context);
    return PsiTreeUtil.getDeepestFirst(el);
  }
}

File: sqlite-android-studio-plugin/src/main/kotlin/com/alecstrong/sqlite/android/util/SqlitePsiUtils.kt
package com.alecstrong.sqlite.android.util

import com.alecstrong.sqlite.android.SQLiteParser
import com.alecstrong.sqlite.android.lang.SqliteLanguage
import com.intellij.lang.ASTNode
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFileFactory
import com.intellij.psi.impl.PsiFileFactoryImpl
import com.intellij.psi.tree.IElementType
import org.antlr.intellij.adaptor.lexer.ElementTypeFactory

object SqlitePsiUtils {
  fun createLeafFromText(project: Project, context: PsiElement?, text: String, type: IElementType) =
      (PsiFileFactory.getInstance(project) as PsiFileFactoryImpl)
          .createElementFromText(text, SqliteLanguage.INSTANCE, type, context)!!.getDeepestFirst()
}

private val RULES = SQLiteParser.ruleNames.asList()
private val TOKENS = SQLiteParser.tokenNames.asList()

internal fun ASTNode.childrenWithRules(vararg rules: Int) = getChildren(
    ElementTypeFactory.createRuleSet(SqliteLanguage.INSTANCE, RULES, *rules))

internal fun ASTNode.childrenWithTokens(vararg tokens: Int) = getChildren(
    ElementTypeFactory.createTokenSet(SqliteLanguage.INSTANCE, TOKENS, *tokens))